AppCenterDataStorage/AppCenterDataStorage/Internal/Client/MSTokenExchange.m

```diff
@@ -1,45 +1,162 @@
 #import "MSTokenExchange.h"
 #import "AppCenter+Internal.h"
+#import "MSDataStorageConstants.h"
 #import "MSDataStoreInternal.h"
+#import "MSKeychainUtil.h"
 #import "MSStorageIngestion.h"
+#import "MSTokenResult.h"
 #import "MSTokensResponse.h"
 
 NS_ASSUME_NONNULL_BEGIN
 
 static NSString *const kMSPartitions = @"partitions";
+static NSString *const kMSTokenResultSucceed = @"Succeed";
+static NSString *const kMSStorageReadOnlyDbTokenKey = @"MSStorageReadOnlyDbToken";
+static NSString *const kMSStorageUserDbTokenKey = @"MSStorageUserDbToken";
 
 @implementation MSTokenExchange : NSObject
 
-+ (void)tokenAsync:(MSStorageIngestion *)httpClient
-           partitions:(NSArray *)partitions
-    completionHandler:(MSGetTokenAsyncCompletionHandler)completion {
++ (void)performDbTokenAsyncOperationWithIngestion:(MSStorageIngestion *)ingestion
+                                        partition:(NSString *)partition
+                                completionHandler:(MSGetTokenAsyncCompletionHandler _Nonnull)completion {
 
-  // Payload.
-  NSError *jsonError;
-  NSData *payloadData = [NSJSONSerialization dataWithJSONObject:@{kMSPartitions : partitions} options:0 error:&jsonError];
+  // Get the cached token if it is saved.
+  MSTokenResult *cachedToken = [MSTokenExchange retrieveCachedToken:partition];
 
-  // Http call.
-  [httpClient sendAsync:payloadData
-      completionHandler:^(NSString *callId, NSHTTPURLResponse *response, NSData *data, NSError *error) {
-        MSLogVerbose([MSDataStore logTag], @"Get token callback, request Id %@ with status code: %lu", callId,
-                     (unsigned long)response.statusCode);
+  // Get a fresh token from the token exchange service if the token is not cached or has expired.
+  if (!cachedToken) {
 
-        // If comletion is provided.
-        if (completion) {
+    // Payload.
+    NSError *jsonError;
+    NSData *payloadData = [NSJSONSerialization dataWithJSONObject:@{kMSPartitions : @[ partition ]} options:0 error:&jsonError];
+
+    // Http call.
+    [ingestion sendAsync:payloadData
+        completionHandler:^(NSString *callId, NSHTTPURLResponse *response, NSData *data, NSError *error) {
+          MSLogVerbose([MSDataStore logTag], @"Get token callback, request Id %@ with status code: %lu", callId,
+                       (unsigned long)response.statusCode);
```

Comment:
`statusCode` is an `NSInteger` not `NSUInteger`, you don't need `unsigned` here.